src/conf/conf_compile.c

```diff
@@ -0,0 +1,517 @@
+
+        if (value->str[1] == 'd') {
+            if (check_type != WT_CONFIG_ITEM_NUM && check_type != WT_CONFIG_ITEM_BOOL)
+                WT_RET_MSG(session, EINVAL, "Value '%.*s' is not compatible with %s type",
+                  (int)value->len, value->str, check->type);
+        } else if (value->str[1] == 's') {
+            if (check_type != WT_CONFIG_ITEM_STRING && check_type != WT_CONFIG_ITEM_STRUCT)
+                WT_RET_MSG(session, EINVAL, "Value '%.*s' is not compatible with %s type",
+                  (int)value->len, value->str, check->type);
+        } else
+            WT_RET_MSG(session, EINVAL, "Value '%.*s' does not match %s for binding",
+              (int)value->len, value->str, "%d or %s");
+
+        bind_offset = top_conf->binding_count++;
+
+        if (conf_key->type == CONF_KEY_BIND_DESC)
+            WT_RET_MSG(session, EINVAL, "Value '%.*s' cannot be used on the same key twice",
+              (int)value->len, value->str);
+
+        conf_key->type = CONF_KEY_BIND_DESC;
+        conf_key->u.bind_desc.type = check_type;
+        conf_key->u.bind_desc.choices = check->choices;
+        conf_key->u.bind_desc.offset = bind_offset;
+        WT_RET(__wt_realloc_def(session, &top_conf->binding_allocated, top_conf->binding_count,
+          &top_conf->binding_descriptions));
+        top_conf->binding_descriptions[bind_offset] = &conf_key->u.bind_desc;
+    } else {
+        switch (check_type) {
+        case WT_CONFIG_ITEM_NUM:
+            if (value->type != WT_CONFIG_ITEM_NUM)
+                WT_RET_MSG(session, EINVAL, "Value '%.*s' expected to be an integer",
+                  (int)value->len, value->str);
+            break;
+        case WT_CONFIG_ITEM_BOOL:
+            if (value->type != WT_CONFIG_ITEM_BOOL &&
+              (value->type != WT_CONFIG_ITEM_NUM || (value->val != 0 && value->val != 1)))
+                WT_RET_MSG(session, EINVAL, "Value '%.*s' expected to be a boolean",
+                  (int)value->len, value->str);
+            break;
+        case WT_CONFIG_ITEM_STRING:
+            /*
+             * Any value passed in, whether it is "123", "true", etc. can be interpreted as a
+             * string. If it must be one of a choice of strings, check that now.
+             */
+            WT_RET(__wt_conf_compile_choice(
+              session, check->choices, value->str, value->len, &value->str));
+            break;
+        case WT_CONFIG_ITEM_ID:
+        case WT_CONFIG_ITEM_STRUCT: /* struct handled previously, needed for picky compilers */
+            return (__wt_illegal_value(session, (int)check_type));
+        }
+
+        conf_key->type = is_default ? CONF_KEY_DEFAULT_ITEM : CONF_KEY_NONDEFAULT_ITEM;
+        conf_key->u.item = *value;
+    }
+    return (0);
+}
+
+/*
+ * __conf_check_compare --
+ *     Compare function used for binary search.
+ */
+static int
+__conf_check_compare(const void *keyvoid, const void *check)
```

Comment:
Having two different types for left and right keys is not safe, as it makes assumptions about the implementation of `bsearch`. (As a point in case, implementations of the equivalent C++ functions do not assume that the right parameter is the "needle" value that you are searching for.)

Reply:
According to the description in https://en.cppreference.com/w/c/algorithm/bsearch, it seems that this behavior is allowed. Hopefully this is indeed a part of the standard! :)

Reply:
Yes, I think this is part of the standard.  I see in
https://pubs.opengroup.org/onlinepubs/9699919799/
```
The comparison function pointed to by compar shall be called with two arguments that point to the key object and to an array element, in that order.
```
I'll add to the comment.